File: problem/easy/findTarget.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        seen = set()
        def dfs(node):
            if not node:
                return False
            # c'est des math k = node1.val + node2.val,
            # donc on cherche node2.val = k - node1.val
            if  k - node.val in seen:
                return True
            seen.add(node.val)
            return dfs(node.left) or dfs(node.right)
        return dfs(root)

File: problem/easy/findTarget.cpp
//
// Created by arnau on 25/06/2024.
//
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

bool findTarget(TreeNode* root, int k) {
    unordered_set<int> s;
    return dfs(root, k, s);
}

bool dfs(TreeNode* root, int k, unordered_set<int>& s) {
    if (root == NULL) return false;
    if (s.count(k - root->val)) return true;
    s.insert(root->val);
    return dfs(root->left, k, s) || dfs(root->right, k, s);
}